File: 10-22-2022/10-22-2022.py
class Node:
    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def serialize(node: Node) -> str:
    if node is None:
        return 'None'
    return f"{node.val} {serialize(node.left)} {serialize(node.right)}"


def deserialize(directories: str) -> Node or None:
    if directories == 'None':
        return None
    directories = directories.split()

    def _deserialize():
        val = directories.pop(0)
        if val == 'None':
            return None
        return Node(val, _deserialize(), _deserialize())

    return _deserialize()


node = Node('root', Node('left', Node('left.left')), Node('right'))
print(serialize(node))
assert deserialize(serialize(node)).left.left.val == 'left.left'
assert node.left.left.val == 'left.left'
assert (deserialize(serialize(node)).right.val == 'right')
assert node.right.val == 'right'

File: 10-22-2022/10-22-2022.js
"use strict";
(() => {
    var _a, _b, _c, _d, _e;
    class Node {
        constructor(val, left = null, right = null) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
    const node = new Node('root', new Node('left', new Node('left.left')), new Node('right'));
    const serialize = (node) => {
        if (node === null)
            return 'X';
        return `${node.val} ${serialize(node.left)} ${serialize(node.right)}`;
    };
    const deserialize = (serialized) => {
        if (serialized === 'X')
            return null;
        const trees = serialized.split(' ');
        const buildTree = (trees) => {
            const val = trees.shift();
            if (val === 'X')
                return null;
            return new Node(val, buildTree(trees), buildTree(trees));
        };
        return buildTree(trees);
    };
    console.log(serialize(node));
    console.log(deserialize(serialize(node)));
    console.log((_b = (_a = node.left) === null || _a === void 0 ? void 0 : _a.left) === null || _b === void 0 ? void 0 : _b.val);
    console.log((_e = (_d = (_c = deserialize(serialize(node))) === null || _c === void 0 ? void 0 : _c.left) === null || _d === void 0 ? void 0 : _d.left) === null || _e === void 0 ? void 0 : _e.val);
})();
